shared/lib/wp/support.js

```diff
@@ -0,0 +1,102 @@
+export default function wpcomSupport( wpcom ) {
+	let supportUser = '';
+	let supportToken = '';
+
+	/**
+	 * Request parameters are as follows:
+	 * - param (required) A string or object that contains the path
+	 * - query (optional) An object that expands the response object
+	 * - body (required*) Required by POST and PUT, but not by GET and DEL
+	 * - fn (required) A callback function to handle the returned result
+	 * - ...rest (optional) Some queries have additional parameters after
+	 *   callback function
+	 *
+	 * Return the index of the query parameter, or if one does not exist,
+	 * return false.
+	 */
+	const getQueryIndex = function( req, args ) {
+		let fnIndex, queryIndex;
+
+		// Find the index of the callback in the arguments
+		fnIndex = args.findIndex( function( e ) {
+			return 'function' === typeof e;
+		} );
+
+		// Set queryIndex based on the request type and fnIndex
+		if ( req === 'post' || req === 'put' ) {
+			queryIndex = ( 3 === fnIndex ) ? 1 : false;
+		} else {
+			queryIndex = ( 2 === fnIndex ) ? 1 : false;
+		}
+		return queryIndex;
+	}
+
+	/**
+	 * Add the supportUser and supportToken to the query.
+	 */
+	const addSupportData = function( query ) {
+		return Object.assign( {}, query, {
+			support_user: supportUser,
+			_support_token: supportToken
+		} );
+	}
+
+	/**
+	 * Mutate the query parameter of the request by adding values for
+	 * support_user and _support_token to the query parameter.
+	 */
+	const extendRequest = function( req, args ) {
+		if ( ! supportUser || ! supportToken ) {
+			return args;
+		}
+
+		let queryIndex = getQueryIndex( req, args );
+		if ( queryIndex ) {
+			args[ queryIndex ] = addSupportData( args[ queryIndex ] );
+		} else {
+			args.splice( 1, 0, addSupportData( {} ) );
+		}
+		return args;
+	}
+
+	const del = wpcom.req.del.bind( wpcom.req );
+	const get = wpcom.req.get.bind( wpcom.req );
+	const post = wpcom.req.post.bind( wpcom.req );
+	const put = wpcom.req.put.bind( wpcom.req );
+
+	return Object.assign( wpcom, {
+		changeUser: function( username, password, fn ) {
+			var args = {
+				apiVersion: '1.1',
+				path: '/internal/support/' + username + '/grant'
+			};
+
+			return wpcom.req.post( args, { password: password }, function( error, response ) {
+				if ( ! error ) {
+					supportUser = response.username;
+					supportToken = response.token;
+				}
+
+				fn( error, response );
+			} );
+		},
+		restoreUser: function() {
+			supportUser = '';
+			supportToken = '';
+		},
+		req: {
+			del: function( ...args ) {
+				return del( ...extendRequest( 'del', args ) );
+			},
+			get: function( ...args ) {
+				return get( ...extendRequest( 'get', args ) );
```

Comment:
You're right, I've been thinking about how to simplify this logic. It's a little tricky to do this though; the `get`, `post`, `put` and `del` functions [can be called in various ways due to optional arguments](https://github.com/Automattic/wpcom.js/blob/master/lib/util/request.js#L41):
```js
// These are equivalent:
wpcom.req.get( params, callback );
wpcom.req.get( params, {}, callback );

// As are these:
wpcom.req.post( params, {}, body, callback );
wpcom.req.post( params, body, callback );
```

Therefore, most of the logic in the `extendRequest` and `getQueryIndex` functions are there in order to determine in which argument the query or body was passed so that we can extend it. I'll think about this some more and see what I can come up with.